### PR TITLE
DevicePool 

### PR DESCRIPTION
**Pull request by aliuTT**
PR for DevicePool. Merging a few concepts in the codebase that is more or less managing multi devices across runtime:
- `ActiveDevices`: tracking some FW builds, device init
- `tt_metal/DeviceMesh`: opens a bunch of devices and returns handles to needed devices. DevicePool will basically be a rename of this.
- `ttnn::device::device_pool::devices`: global handle for all devices, this will be deleted

I'm also trying out something that isn't really tested in backend:
```
Device *dev = new Device();
// run some tests
dev->close();
dev->initialize();
// run some tests
```
I'll also clean up the instance method to be something like this:
```
tt::DevicePool::initialize(device_ids, num_hw_cqs, l1_small_size);
tt::DevicePool::instance().get_all_devices();
```

### CODE REVIEW

File: tt_metal/impl/device/device_pool.hpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    static void initialize(
+        std::vector<chip_id_t> device_ids,
+        const uint8_t num_hw_cqs,
+        size_t l1_small_size,
+        const std::vector<uint32_t> &l1_bank_remap = {}) noexcept {
+      std::cout << " DP initialize " << std::endl;
+      std::cout << " device ids ";
+      for (const auto &chip: device_ids) { std::cout << chip << " " << std::endl; }
+      std::cout << " num hw cqs " << num_hw_cqs << " l1 small " << l1_small_size << std::endl;
```
Comment (aliuTT) [inline tt_metal/impl/device/device_pool.hpp:35]: will remove prints

File: tt_metal/impl/device/device_pool.cpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+       this->devices.resize(id + 1);
+    }
+    if (this->devices[id] == nullptr) {
+        auto dev = new Device(id, this->num_hw_cqs, this->l1_small_size, this->l1_bank_remap);
+        dev->build_firmware();
+        this->devices[id] = std::unique_ptr<Device>(dev);
+
+    } else {
+      const auto& dev = this->devices[id];
+      std::cout << " DP re-init device " << id << std::endl;
```
Comment (aliuTT) [inline tt_metal/impl/device/device_pool.cpp:45]: will remove prints

File: tests/ttnn/unit_tests/operations/test_conv2d.py [hunk 3]
```diff
@@ -373,6 +373,7 @@ def test_resnet50_conv_gs(
     pad_w,
     use_1d_systolic_array,
 ):
+    pytest.skip("Skipping, suspecting some device state dependency")
```
Comment (xanderchin) [inline tests/ttnn/unit_tests/operations/test_conv2d.py:376]: Is there an issue for this being disabled?

File: tt_metal/impl/device/device.cpp [hunk 4]
```diff
@@ -179,6 +136,7 @@ void Device::initialize_build() {
 }
 
 void Device::build_firmware() {
+  std::cout << " BUILDING FIRMWARE " << this->id() << std::endl;
```
Comment (xanderchin) [inline tt_metal/impl/device/device.cpp:139]: Logger?

File: tt_metal/impl/device/device_pool.hpp (hunk 1 above)
Comment (tt-rkim) [reply tt_metal/impl/device/device_pool.hpp:35]: thanks, pls re-ping for approval once done

**Review by cfjchu — APPROVED** [review]

**Review by mo-tenstorrent — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/test_conv2d.py (hunk 3 above)
Comment (cfjchu) [reply tests/ttnn/unit_tests/operations/test_conv2d.py:376]: make sure these issues are resolved before merge

File: tt_metal/impl/device/device_pool.cpp (hunk 2 above)
Comment (aliuTT) [reply tt_metal/impl/device/device_pool.cpp:45]: removed prints

File: tests/ttnn/unit_tests/operations/test_conv2d.py (hunk 3 above)
Comment (aliuTT) [reply tests/ttnn/unit_tests/operations/test_conv2d.py:376]: got rid of all the pytest skips, ended up fixing all the segfaults instead 🙃

File: tt_metal/detail/tt_metal.hpp [hunk 5]
```diff
@@ -39,13 +41,13 @@ namespace tt::tt_metal{
         }
 
         std::map<chip_id_t, Device *> CreateDevices(
+            // TODO: delete this in favour of DevicePool
```
[the remark this answers is not in the document]
Comment (aliuTT) [reply tt_metal/detail/tt_metal.hpp:44]: I think we actually keep this around. It's a good separation from DevicePool's active devices, to user devices aliased via ids.

File: tt_metal/detail/tt_metal.hpp [hunk 6]
```diff
@@ -155,6 +157,8 @@ namespace tt::tt_metal{
          * | device        | The device holding the program being profiled.    | Device *        |                           | True     |
          * */
 	    void InitDeviceProfiler(Device *device);
+      // API to clear all device profilers
+	    void ClearDeviceProfiler();
```
[the remark this answers is not in the document]
Comment (aliuTT) [reply tt_metal/detail/tt_metal.hpp:161]: Removed this api, I added to main a different way to deallocate buffers on device close

File: tt_metal/impl/device/device.cpp [hunk 7]
```diff
@@ -1325,7 +1261,8 @@ bool Device::close() {
                         not_done_dispatch_cores.insert(phys_core);
                         log_debug(tt::LogMetal, "MMIO Device Prefetch core: Logical: {} - Physical: {}", prefetch_location.str(), phys_core.str());
                     }
-                } else if (this->active_devices_.is_device_active(device_id)) {
+                } else  {
+                    // Assume that R chip is always open
```
[the remark this answers is not in the document]
Comment (aliuTT) [reply tt_metal/impl/device/device.cpp:1265]: I think this was a temp change that made it here. I've reverted.

**Review by abhullar-tt — APPROVED** [review]

**Comment by aliuTT** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/9469308539

**Review by xanderchin — APPROVED** [review]

**Review by TT-BrianLiu — APPROVED** [review]

**Review by tt-asaigal — APPROVED** [review]

**Pull request merged**